src/mbgl/tile/geometry_tile.cpp

```diff
@@ -93,12 +93,13 @@ void GeometryTile::setLayers(const std::vector<Immutable<Layer::Impl>>& layers)
     pending = true;
 
     std::vector<Immutable<Layer::Impl>> impls;
+    impls.reserve(layers.size());
 
     for (const auto& layer : layers) {
         // Skip irrelevant layers.
-        if (layer->getTypeInfo()->source == LayerTypeInfo::Source::NotRequired ||
-            layer->source != sourceID ||
-            id.overscaledZ < std::floor(layer->minZoom) ||
+        assert(layer->getTypeInfo()->source != LayerTypeInfo::Source::NotRequired);
+        assert(layer->source == sourceID);
+        if (id.overscaledZ < std::floor(layer->minZoom) ||
             id.overscaledZ >= std::ceil(layer->maxZoom) ||
             layer->visibility == VisibilityType::None) {
```

Comment:
nit: visibility is checked in `RenderLayer::needsRendering`, might be converted to assert as well?

Reply:
`needsRendering == true` for a source if there is at least one visible layer, but all the others are also passed -> needs checking :disappointed:  (though the check might be optimized out in future).